planning/autoware_freespace_planning_algorithms/src/astar_search.cpp

```diff
@@ -333,46 +333,60 @@ void AstarSearch::setPath(const AstarNode & goal_node)
   header.stamp = rclcpp::Clock(RCL_ROS_TIME).now();
   header.frame_id = costmap_.header.frame_id;
 
-  waypoints_.header = header;
-  waypoints_.waypoints.clear();
-
   // From the goal node to the start node
   const AstarNode * node = &goal_node;
 
+  std::vector<PlannerWaypoint> waypoints;
+
   geometry_msgs::msg::PoseStamped pose;
   pose.header = header;
 
-  const auto interpolate = [this, &pose](const AstarNode & node) {
+  const auto interpolate = [this, &waypoints, &pose](const AstarNode & node) {
     if (node.parent == nullptr || !astar_param_.adapt_expansion_distance) return;
     const auto parent_pose = node2pose(*node.parent);
     const double distance_2d = calcDistance2d(node2pose(node), parent_pose);
     const int n = static_cast<int>(distance_2d / min_expansion_dist_);
     for (int i = 1; i < n; ++i) {
-      const double dist = ((distance_2d * i) / n) * (node.is_back ? -1.0 : 1.0);
+      double dist = ((distance_2d * i) / n) * (node.is_back ? -1.0 : 1.0);
+      if (is_backward_search_) dist *= -1.0;
       const double steering = node.steering_index * steering_resolution_;
       const auto local_pose = kinematic_bicycle_model::getPose(
         parent_pose, collision_vehicle_shape_.base_length, steering, dist);
       pose.pose = local2global(costmap_, local_pose);
-      waypoints_.waypoints.push_back({pose, node.is_back});
+      waypoints.push_back({pose, node.is_back});
     }
   };
 
   // push astar nodes poses
   while (node != nullptr) {
     pose.pose = local2global(costmap_, node2pose(*node));
-    waypoints_.waypoints.push_back({pose, node->is_back});
+    waypoints.push_back({pose, node->is_back});
     interpolate(*node);
     // To the next node
     node = node->parent;
   }
 
```

Comment:
Maybe add some check to make sure `waypoints` is not empty here. Or even to make sure it has at least 2 elements, not sure. Some operations below here depend on this.